feat: add placeholderValue to page source tree

ikharebashviliGD opened the pull request:
**Summary**
This PR adds support for the placeholderValue attribute to be included in the response from the `/source` (`getPageSource`) endpoint.

**Motivation**
Previously, `placeholderValue` was only available when requesting attributes for a specific element. However, this attribute is also valuable when analyzing the full accessibility tree returned by `/source`, particularly for identifying input fields and verifying placeholder text during automated accessibility testing.

By including `placeholderValue` in the full snapshot, we enable better support for end-to-end validations and tools that rely on the complete accessibility hierarchy.

ikharebashviliGD committed:
Include placeholderValue in /source endpoint response

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -216,13 +217,18 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
           return [@([wrappedSnapshot isWDFocused]) stringValue];
+      },
+      FBExclusionAttributePlaceholderValue: ^{
+          return FBValueOrNull(wrappedSnapshot.wdPlaceholderValue);
```
I'm not sure if it makes sense to have nulls everywhere as this attribute only makes sense for a limited range of elements

ikharebashviliGD replied:
> I'm not sure if it makes sense to have nulls everywhere as this attribute only makes sense for a limited range of elements

Thanks for the feedback — you're absolutely right that `placeholderValue` only applies to a limited set of elements.
I've updated the implementation to include it only for relevant types (`TextField`, `SecureTextField`, `SearchField`) and excluded it entirely from the output for all others.
This ensures the `/source` response stays clean and focused, without unnecessary `null` values. Let me know if you'd prefer a different approach!

ikharebashviliGD committed:
feat: add placeholderValue to /source for text input elements only

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -201,7 +202,8 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
   
-  NSDictionary<NSString *, NSString * (^)(void)> *attributeBlocks = @{
+  // Define the base attribute blocks that apply to all elements.
+  NSDictionary<NSString *, NSString * (^)(void)> *baseAttributeBlocks = @{
```
I think it makes sense to extract this logic into a separate method for better readability.

Also this dictionary should be defined as mutable to reduce the amount of boilerplate code

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -218,11 +220,33 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
+  
+  // Add placeholderValue only for elements where it is meaningful (e.g., text input fields).
+  switch (snapshot.elementType) {
+      case XCUIElementTypeTextField:
```
consider reusing this helper for detecting whether a particular element is a text field: https://github.com/appium/WebDriverAgent/blob/9c23d91c794a1ce163799918596968b2968939ed/WebDriverAgentLib/Categories/XCUIElement%2BFBWebDriverAttributes.m#L73

ikharebashviliGD replied:
@mykola-mokhnach I initially tried to use `fb_supportsInnerText` as suggested, but it's only available on `XCUIElement`, while here we work exclusively with `FBXCElementSnapshotWrapper`.
Rather than casting or duplicating the logic inline, I added a minimal helper `fb_supportsPlaceholder` to the existing `FBXCElementSnapshotWrapper+Helpers` category.
This keeps everything localized, avoids unnecessary switching, and makes it easier to extend later.

KazuCocoa replied:
Would exposing `fb_supportsInnerText` from `XCUIElement+FBWebDriverAttributes.h` help to refer to it?

mykola-mokhnach replied:
I've extracted this helper in the other [PR](https://github.com/appium/WebDriverAgent/pull/1018). Please rebase with master to be able to use it

ikharebashviliGD committed:
refactor: extract attribute block building logic into helper method

Moved the logic for constructing attribute blocks into a separate method
for clarity and future extensibility. Also replaced direct type checks
with a reusable fb_supportsPlaceholder helper on FBXCElementSnapshotWrapper.

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -248,6 +237,41 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
 
+// Private helper that builds all attribute blocks for a given snapshot.
+// Includes both base attributes and any element-specific ones (e.g. placeholder for text inputs, etc.).
++ (NSMutableDictionary<NSString *, NSString *(^)(void)> *)attributeBlockMapForSnapshot:(id<FBXCElementSnapshot>)snapshot
```
The returned dictionary must be immutable

mykola-mokhnach replied:
method names must always start with `fb_` in extension classes

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -248,6 +237,41 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
   return info;
 }
 
+// Private helper that builds all attribute blocks for a given snapshot.
```
consider moving the helper closer to the tail of the class definition if it is private

ikharebashviliGD replied:
> consider moving the helper closer to the tail of the class definition if it is private

Updated the comment to clarify that this helper is tightly coupled with dictionaryForElement: and placed intentionally right below it for clarity.

ikharebashviliGD committed:
refactor: use FBDoesElementSupportInnerText and clarify helper method comment

- Replaced custom placeholder support check with FBDoesElementSupportInnerText, as introduced in upstream PR
- Updated comment for fb_attributeBlockMapForSnapshot to clarify usage context and avoid confusion about method placement

mykola-mokhnach commented on WebDriverAgentLib/Utilities/FBElementHelpers.m:
```diff
@@ -0,0 +1,17 @@
+
+#import "FBElementHelpers.h"
+
+BOOL FBDoesElementSupportInnerText(XCUIElementType elementType) {
```
I believe these changes should not be visible as they are already part of the master branch. Was the rebase successful?

ikharebashviliGD replied:
@mykola-mokhnach Fixed

mykola-mokhnach commented on WebDriverAgentLib/Categories/XCUIApplication+FBHelpers.m:
```diff
@@ -248,6 +238,43 @@ + (NSDictionary *)dictionaryForElement:(id<FBXCElementSnapshot>)snapshot
 
+// Helper used by `dictionaryForElement:` to assemble attribute value blocks,
+// including both common attributes and conditionally included ones like placeholderValue.
++ (NSDictionary<NSString *, NSString *(^)(void)> *)fb_attributeBlockMapForSnapshot:(id<FBXCElementSnapshot>)snapshot
```
we do not use the `snapshot` instance in this method. Also `wrappedSnapshot` anyway contains it

ikharebashviliGD replied:
@mykola-mokhnach Totally agree, removed the unused `snapshot` parameter and renamed the method. Also noticed some tests are failing — should I do something on my side or can someone help with that?

mykola-mokhnach replied:
some tests are unstable due to CI slowness. I've verified their results manually

ikharebashviliGD committed:
refactor: remove extra parameter from the fb_attributeBlockMapForWrappedSnapshot method

KazuCocoa commented:
Hi @ikharebashviliGD, congrats: the Appium project wants to compensate you for this contribution! Please reply to this comment mentioning @jlipps and @KazuCocoa and sharing your [OpenCollective](https://opencollective.com/) account name, so that we can initiate payment! Or let us know if you decline to receive compensation via OpenCollective. Thank you!

ikharebashviliGD commented:
> Hi @ikharebashviliGD, congrats: the Appium project wants to compensate you for this contribution! Please reply to this comment mentioning @jlipps and @KazuCocoa and sharing your [OpenCollective](https://opencollective.com/) account name, so that we can initiate payment! Or let us know if you decline to receive compensation via OpenCollective. Thank you!

@jlipps @KazuCocoa Thank you very much! It's a pleasure.
Here is my OpenCollective account: https://opencollective.com/illia-kharebashvili
